Endpoint healthz reported via command & in endpoint get

raybejjani opened the pull request:
**Summary of changes**:
Added healthz to `endpoint get` and introduced `endpoint healthz`. It pretty prints and returns json

@tgraf I guessed on some of this logic, let me know if it is incorrect.

Sample output:
```
$ cilium endpoint healthz 175 --json
{
  "bpf": "OK",
  "connected": true,
  "health": "OK",
  "policy": "OK"
}

$ cilium endpoint healthz 175
Health:      OK
BPF:         OK
Policy:      OK
Connected:   yes
```

Fixes: #1277 

```release-note
Added cilium endpoint healthz command that returns a summary of the endpoint's health
```

**How to test (optional)**:
Tests pass.
make clean && make generate-api && make reload
Note: There might be new/old files left in the api directory. In that case:
rm -rf ./api/ && git checkout -f

aanm commented on pkg/endpoint/endpoint.go:
```diff
@@ -449,6 +450,101 @@ func (e *Endpoint) GetModelRLocked() *models.Endpoint {
+	}
+
+	policyStatus := models.EndpointHealthzStatusOK
+	switch currentState {
```
Why do you have a duplicate switch? You can assign `policyStatus` and `bpfStatus` at the same time in a single switch

raybejjani replied:
Yeah, this was a bad idea. I expected more variation. I collapsed everything into a single switch.

aanm commented on pkg/endpoint/endpoint.go:
```diff
@@ -449,6 +450,101 @@ func (e *Endpoint) GetModelRLocked() *models.Endpoint {
+		policyStatus = models.EndpointHealthzStatusBootstrap
+
+	case models.EndpointStateNotReady:
+		bpfStatus = models.EndpointHealthzStatusWarning
```
`s/bpfStatus/policyStatus`

aanm commented on pkg/endpoint/endpoint.go:
```diff
@@ -449,6 +450,101 @@ func (e *Endpoint) GetModelRLocked() *models.Endpoint {
+		bpfStatus = models.EndpointHealthzStatusWarning
+
+	case models.EndpointStateDisconnected:
+		bpfStatus = models.EndpointHealthzStatusDisabled
```
`s/bpfStatus/policyStatus`

aanm commented on pkg/client/endpoint.go:
```diff
@@ -74,6 +74,16 @@ func (c *Client) EndpointLogGet(id string) (models.EndpointStatusLog, error) {
 	return resp.Payload, nil
 }
 
+// EndpointLogHealthz returns endpoint healthz
```
`s/EndpointLogHealthz/EndpointHealthzGet`

aanm commented on pkg/endpoint/endpoint.go:
```diff
@@ -453,7 +453,7 @@ func (e *Endpoint) GetModelRLocked() *models.Endpoint {
 //
 // Must be called with e.Mutex locked.
-func (e *Endpoint) getHealthzModel() *models.EndpointHealthz {
+func (e *Endpoint) GetHealthzModel() *models.EndpointHealthz {
```
Create a public method that can be called without the e.Mutex locked. You are calling this function without locking the endpoint in daemon/endpoint.go:581

raybejjani replied:
You are correct. fixed.

tgraf commented on api/v1/openapi.yaml:
```diff
@@ -755,6 +755,9 @@ definitions:
       status:
         description: Most recent status log. See endpoint/{id}/log for the complete log.
         "$ref": "#/definitions/EndpointStatusLog"
+      healthz:
```
I'm all for using `healthz` in the API as it has become a convention but this looks ugly for internal member names.

tgraf commented on api/v1/openapi.yaml:
```diff
@@ -835,6 +838,29 @@ definitions:
+    description: Health of the endpoint
+    type: object
+    properties:
+      health:
```
`overallHealth`?

raybejjani replied:
done

tgraf commented on api/v1/openapi.yaml:
```diff
@@ -835,6 +838,29 @@ definitions:
+        description: Is this endpoint reachable
+        type: boolean
+  EndpointHealthzStatus:
+    description: A common set of statuses for endpoint health
```
Needs an explanation of the different states

raybejjani replied:
I'm not sure that I did it correctly, there doesn't seem to be an official way to do this (https://swagger.io/docs/specification/2-0/enums/ ?)

tgraf commented on api/v1/openapi.yaml:
```diff
@@ -835,6 +838,29 @@ definitions:
+    enum:
+      - OK
+      - Bootstrap
+      - Pending
```
Thinking more about this, I think we want to be a bit more fine grained and split this into:
- ChangesPending
- CreationPending

raybejjani replied:
Isn't bootstrap CreationPending? I can change the name on that.

tgraf replied:
OK, I'm fine with that then

raybejjani committed:
cli: Add healthz field to cilium status & endpoint health

We now report a status for endpoints, under health. It reflects a
top-level view of whether the endpoint is enforcing policy.

Signed-off-by: Ray Bejjani <ray@covalent.io>
